dmscripts/compare_dmp_and_dnb_data.py

```diff
@@ -0,0 +1,324 @@
+from dmutils.env_helpers import get_api_endpoint_from_stage
+
+
+def levenshtein_ratio_and_distance(s, t, ratio_calc=False):
+    """ levenshtein_ratio_and_distance:
+        Calculates levenshtein distance between two strings.
+        If ratio_calc = True, the function computes the
+        levenshtein distance ratio of similarity between two strings
+        For all i and j, distance[i,j] will contain the Levenshtein
+        distance between the first i characters of s and the
+        first j characters of t
+    """
+    # Initialize matrix of zeros
+    rows = len(s) + 1
+    cols = len(t) + 1
+    distance = np.zeros((rows, cols), dtype=int)
+
+    # Populate matrix of zeros with the indeces of each character of both strings
+    for i in range(1, rows):
+        for k in range(1, cols):
+            distance[i][0] = i
+            distance[0][k] = k
+
+    # Iterate over the matrix to compute the cost of deletions,insertions and/or substitutions
+    for col in range(1, cols):
+        for row in range(1, rows):
+            if s[row - 1] == t[col - 1]:
+                cost = 0  # If the characters are the same in both strings in a given position [i,j] then the cost is 0
+            else:
+                # In order to align the results with those of the Python Levenshtein package,
+                # if we choose to calculate the ratio the cost of a substitution is 2.
+                # If we calculate just distance, then the cost of a substitution is 1.
+                if ratio_calc is True:
+                    cost = 2
+                else:
+                    cost = 1
+            distance[row][col] = min(
+                distance[row - 1][col] + 1,  # Cost of deletions
+                distance[row][col - 1] + 1,  # Cost of insertions
+                distance[row - 1][col - 1] + cost  # Cost of substitutions
+            )
+    if ratio_calc is True:
+        # Computation of the Levenshtein Distance Ratio
+        Ratio = ((len(s) + len(t)) - distance[row][col]) / (len(s) + len(t))
+        return Ratio
+    else:
+        # print(distance) # Uncomment to see the matrix showing how the algorithm computes the cost of deletions,
+        # insertions and/or substitutions
+        # This is the minimum number of edits needed to convert string a to string b
+        return f"The strings are {distance[row][col]} edits away"
+
+
+class DNBAPIException(Exception):
+    "Exception class for the D&B API errors, list here: https://docs.dnb.com/direct/2.0/en-US/response-codes"
+    def __init__(self, code, text):
+        self.code = code
+        self.text = text
+
+
+class DNBDirectAPIClient(object):
+    "A client utilising the D&B direct API with methods used by the DMp API."
+
+    def __init__(self, username, password, stage='production'):
+        "initialisation with D&B credentials and optionally DMp stage"
+        self.stage = stage
+        self.data_api_client = None
```

Comment:
Are we supposed to just assign this `data_api_client` to the instance after construction? Would be nicer as an `__init__` arg...

Reply:
Oh no I see it tries to autogenerate itself when needed...

Reply:
> Oh no I see it tries to autogenerate itself when needed...

Yes, and in addition the client has methods that independent of our API, only concerned with the D&B API.